### PR TITLE
Ogmios package, module and test

### LINKED ISSUE #30: NixOS package for `ogmios`
Refer to [ogmios-nixos](https://github.com/mlabs-haskell/ogmios-nixos), ideally that repository will be replaced by `cardano.nix` when this issue is implemented.

Ogmios doesn't provide a derivation upstream, so it has to live here.
It would be nice if this flake would also expose that derivation between its `packages` output. Probably we want to invent some machinery like we did for tests and modules to automatically read derivations from paths like `packages/ogmios/default.nix` so that when we'll add another package it we can just add the derivation in `packages/<name>/default.nix`.

Probably this will give us problems for non `x86_64-linux` system, at the moment I would say to not waste too much time trying to make it work for other architectures. If it doesn't work for other architectures it would be nice having a smart way to mark it as broken without making Hercules CI fail when trying to build the `packages` outputs for all the systems (trivially we can make `flake-parts` avoid to expose those derivations for those systems if marked as broken).

Feel free to split this issue into sub issues, in case please link them here as a to do list.

### PR DESCRIPTION
**Pull request by brainrake**
fixes #30 #31 #32
- cardano-node 8.8.* and 8.9.* do not build so we're stuck with 8.7.3
- work around cardano-node socket permissions issue: poll and chmod
- work around cardano-node configuration json sharing issue: set from cardano-configurations for all services (found specific version for cardano-node 8.7.3)
- add ogmios binary release ~6.1.0~ 6.2.0 because it fails to build with haskell.nix and 6.0.3 is too old
- add ogmios module from ogmios-nixos
- update ogmios module to work with newer ogmios
- update ogmios module to work with cardano.nix cardano-node
- avoid segfaults, lock down ogmios systemd security
- update module field documentation
- update vmTests (formerly nixosTests) to automatically handle checks, effects, and add apps to run tests
- update vmTests to expose check and effect derivations
- fix run-vm-test argument quotation and shellcheck errors
- add vm test to check ogmios sync progress
- update readme, detail how to run integration test showing cardano-node and ogmios works

### CODE REVIEW

File: flake.nix [hunk 1]
```diff
@@ -1,36 +1,32 @@
...
 
+    # cardano-node
     cardano-node.url = "github:intersectmbo/cardano-node?ref=8.7.3";
+    cardano-configurations.url = "github:input-output-hk/cardano-configurations/21249e0d5c68b4e8f3661b250aa8272a8785d678";
+    cardano-configurations.flake = false;
 
     # Utilities
-    devshell = {
-      url = "github:numtide/devshell";
-      inputs.nixpkgs.follows = "nixpkgs";
```
Comment (avnik) [inline flake.nix:14]: Any reason to throw out nixpkgs override here? Downloading extra nixpkgs slow-down first build significantly

Comment (avnik) [reply flake.nix:14]: Ahh, I see -- you group `follows` statement differently, but is pretty counterintuitive (for me personally)

File: modules/globals.nix [hunk 2]
```diff
@@ -11,22 +11,19 @@ in
...
-        description = ''
-          Cardano network number to join/use (should match cardanoNix.globals,network)
-        '';
-        internal = true;
       };
       networkNumbers = lib.mkOption {
+        description = "Map from network names to network numbers. Selected network must be present in the map";
         type = types.attrsOf types.int;
         default = {
           mainnet = 0;
```
Comment (avnik) [inline modules/globals.nix:29]: FYI,  mainnet is not zero. That actually a stub added for scaffolding, we need fetch ids from caradno-configuration, iohk.nix or whatever.  (not blocking for your PR, but if you do any amendament -- add a link to the issue https://github.com/mlabs-haskell/cardano.nix/issues/53, but if you decide to fix it -- add closing comment for it)

File: modules/ogmios.nix [hunk 3]
```diff
@@ -0,0 +1,25 @@
+{
+  config,
+  lib,
+  ...
+}: let
+  cfg = config.cardanoNix.ogmios;
+in {
+  options.cardanoNix.ogmios = {
+    enable = lib.mkEnableOption "cardano-node service";
```
Comment (avnik) [inline modules/ogmios.nix:9]: ogmios service, not cardano-node

File: packages/ogmios.nix [hunk 4]
```diff
@@ -0,0 +1,34 @@
+{
+  perSystem = {
+    system,
+    pkgs,
+    ...
+  }: {
+    packages.ogmios = pkgs.stdenv.mkDerivation rec {
+      pname = "ogmios";
```
Comment (avnik) [inline packages/ogmios.nix:8]: Add an issue about "ogmios doesn't build from source". I'll probably try to fix it, if time budget allows.

File: tests/cardano-node.nix [hunk 5]
```diff
@@ -24,7 +22,8 @@
 
       testScript = ''
         machine.wait_for_unit("cardano-node")
-        machine.wait_until_succeeds("""[[ $(echo "$(cardano-cli query tip --testnet-magic 2 | jq '.syncProgress' --raw-output) > 0.01" | bc) == "1" ]]""")
+        machine.wait_until_succeeds("""[[ $(echo "$(cardano-cli query tip --testnet-magic 2 | jq '.syncProgress' --raw-output) > 0.001" | bc) == "1" ]]""")
```
Comment (avnik) [inline tests/cardano-node.nix:25]: Hardcoded testnet magic. We should use current `cardano.nix` default magic, if specified (and unless it a mainnet)

**Review by avnik — DISMISSED** [review]
Generally all comments are non-blocking,  if you argument -- I can withdraw my comment, and approve during today meeting.
General LGTM

File: flake.nix (hunk 1 above)
Comment (brainrake) [reply flake.nix:14]: :+1:

File: packages/ogmios.nix (hunk 4 above)
Comment (brainrake) [reply packages/ogmios.nix:8]: :+1:
I saw a commit on master that should fix haskell.nix build. So let's wait until next release.

File: tests/cardano-node.nix (hunk 5 above)
Comment (brainrake) [reply tests/cardano-node.nix]: Fixed to get magic from cardanoNix settings.

File: packages/ogmios.nix (hunk 4 above)
Comment (brainrake) [reply packages/ogmios.nix:9]: Added comment.

File: modules/ogmios.nix (hunk 3 above)
Comment (brainrake) [reply modules/ogmios.nix]: Fixed.

File: modules/globals.nix (hunk 2 above)
Comment (brainrake) [reply modules/globals.nix]: Added correct id.

File: flake.nix (hunk 1 above)
Comment (brainrake) [reply flake.nix:14]: Reorganized to normal layout.

File: README.md [hunk 6]
```diff
@@ -23,23 +23,46 @@ Don't edit `~/.config/nix/nix.conf` in your home directory. Don't add users to `
...
+
+  fmt                     - format the source tree
+  update-pre-commit-hooks - update git pre-commit hooks
 ```
 
 A `.envrc` is also provided, using [direnv]() and [nix-direnv](https://github.com/nix-community/nix-direnv) is suggested.
 
+### Run Integration Test
+
+From the devshell, run an integration test that starts `cardano-node` and `ogmios` on the `preview` testnet and checks for synchronization progress.
```
Comment (aciceri) [inline README.md:60]:
```suggestion
### Running Integration Tests

From the devshell you can run integration tests with `run-vm-test`, for example the following will start `cardano-node` and `ogmios` on the `preview` testnet and will check for synchronization progress.
```

Comment (aciceri) [reply README.md:60]: Just wanted to make it a bit more precise, feel free to rephrase (or change again the title, I'm not convinced)

File: checks/run-vm-test.nix [hunk 7]
```diff
@@ -6,29 +6,29 @@
   ...
 }:
 writeShellApplication {
-  name = "run-test";
+  name = "run-vm-test";
```
Comment (aciceri) [inline checks/run-vm-test.nix:9]: Why you didn't like `run-test`? `run-vm-test` seems to imply that it's only one vm while it can run multiple ones.

File: ci/default.nix [hunk 8]
```diff
@@ -22,7 +22,6 @@
     perSystem = {config, ...}: {
       hercules-ci.github-pages.settings.contents = config.packages.docs;
     };
-    herculesCI.ciSystems = ["x86_64-linux" "x86_64-darwin"];
```
Comment (aciceri) [inline ci/default.nix:25]: Why did you remove this?

File: modules/default.nix
Comment (aciceri) [inline modules/default.nix:1]: Yeah probably exporting modules adding overlays like this is the saner thing to do in the end...

File: modules/services/ogmios.nix [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+        RestrictAddressFamilies = ["AF_UNIX" "AF_INET" "AF_INET6"];
+        RestrictNamespaces = true;
+        LockPersonality = true;
+        RestrictRealtime = true;
+        RestrictSUIDSGID = true;
+        RemoveIPC = true;
+        PrivateMounts = true;
+        SystemCallArchitectures = "native";
+        SystemCallFilter = ["@system-service" "~@privileged"];
+        MemoryDenyWriteExecute = true;
```
Comment (aciceri) [inline modules/services/ogmios.nix:114]: Where did you take these options from? It's not the first time I see them, they seem to be good defaults, perhaps we should consider put them in a different place if they are going to be used by all the modules we'll define.

File: packages/cardano.nix [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+  inputs,
+  lib,
+  ...
+}: {
+  perSystem = {system, ...}: {
+    packages = lib.filterAttrs (_: v: v != null) {
+      cardano-cli = inputs.cardano-node.packages.${system}.cardano-cli or null;
+      cardano-node = inputs.cardano-node.packages.${system}.cardano-node or null;
+      inherit (inputs) cardano-configurations;
+    };
```
Comment (aciceri) [inline packages/cardano.nix:11]:
```suggestion
  perSystem = {inputs', ...}: {
    packages = lib.filterAttrs (_: v: v != null) {
      cardano-cli = inputs'.cardano-node.packages.cardano-cli or null;
      cardano-node = inputs'.cardano-node.packages.cardano-node or null;
      inherit (inputs') cardano-configurations;
    };
```

File: packages/default.nix
Comment (aciceri) [inline packages/default.nix:1]: So this creates an overlay for each package? I believe that when for a big `n` having 1 overlay with `n` packages is way faster than `n` overlays with 1 package each.

Anyway this was just a theoretical consideration about something which is not a problem yet (if it will ever be). Ignore it.

File: packages/ogmios.nix (hunk 4 above)
Comment (aciceri) [reply packages/ogmios.nix:9]: Do you know how often they release?

File: tests/ogmios.nix [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+
+      testScript = {nodes, ...}: let
+        magic = toString nodes.machine.config.cardanoNix.globals.networkNumber;
+      in ''
+        machine.wait_for_unit("cardano-node")
+        machine.wait_for_unit("cardano-node-socket")
+        machine.wait_until_succeeds("""[[ $(echo "$(cardano-cli query tip --testnet-magic ${magic} | jq '.syncProgress' --raw-output) > 0.001" | bc) == "1" ]]""")
+        machine.wait_for_unit("ogmios")
+        machine.succeed("curl --fail http://localhost:1337")
+        machine.wait_until_succeeds(r"""journalctl --no-pager -r -n 1 -u ogmios.service -g networkSynchronization\":0\.[0-9][0-9][0-9][1-9]""")
```
Comment (aciceri) [inline tests/ogmios.nix:30]: If you want `ogmios` also exposes `/health` with some information we could check from the test

**Comment by aciceri** [comment]
Oh also the documentation is not generated for modules in `modules/services`

File: modules/services/ogmios.nix (hunk 9 above)
Comment (brainrake) [reply modules/services/ogmios.nix:114]: `systemd-analyze security ogmios.service`
https://hercules-ci.com/github/mlabs-haskell/cardano.nix/jobs/345/effects/effects.vmTests-ogmios/log
each service needs different things so it might not make sense to make a module

File: checks/run-vm-test.nix (hunk 7 above)
Comment (brainrake) [reply checks/run-vm-test.nix:9]: To have the same naming as the module.
I kept it singular because it can only run one test.

File: ci/default.nix (hunk 8 above)
Comment (brainrake) [reply ci/default.nix:25]: It can build all the systems in flake.systems I think, so we don't need to shrink it down. Not sure what the intention was of having it here separately. Maybe we want stuff in the flake that ci can't currently build?

File: packages/default.nix
Comment (brainrake) [reply packages/default.nix:1]: I don't think it matters since the packages in the overlay don't depend on other packages, and there are only a few of them. I wanted to group similar packages together but right now each group has only one package.

File: packages/cardano.nix (hunk 10 above)
Comment (brainrake) [reply packages/cardano.nix:11]: unfortunately this doesn't work because `transpose` tries to evaluate systems that are not in the input, so reverted

File: packages/default.nix
Comment (brainrake) [reply packages/default.nix:1]: it's a very small number and the packages don't depend on the other packges in the set, so it should be fine

**Comment by brainrake** [comment]
Added issue for documentation since I could not fix it quickly.
https://github.com/mlabs-haskell/cardano.nix/issues/54

**Comment by brainrake** [comment]
Resolved all issues discussed above.

**Comment by brainrake** [comment]
ogmios 6.2.0 just came out

File: modules/services/ogmios.nix (hunk 9 above)
Comment (aciceri) [reply modules/services/ogmios.nix:114]: In practice how do you choose them? Do you apply them in the stricter way possible and then manually check if they all makes sense? And then test if it works?

Is there a list with all the security options?

**Review by aciceri — APPROVED** [review]
